hardware_interface/include/hardware_interface/actuator_interface.hpp

```diff
@@ -115,33 +120,202 @@ class ActuatorInterface : public rclcpp_lifecycle::node_interfaces::LifecycleNod
+    {
+      joint_state_interfaces_.insert(std::make_pair(description.get_name(), description));
+    }
+  }
+
+  /**
+   * Import the InterfaceDescription for the CommandInterfaces from the HardwareInfo.
+   * Separate them into the possible types: Joint and store them.
+   */
+  virtual void import_command_interface_descriptions(const HardwareInfo & hardware_info)
+  {
+    auto joint_command_interface_descriptions =
+      parse_command_interface_descriptions(hardware_info.joints);
+    for (const auto & description : joint_command_interface_descriptions)
+    {
+      joint_command_interfaces_.insert(std::make_pair(description.get_name(), description));
+    }
+  }
+
   /// Exports all state interfaces for this hardware interface.
   /**
-   * The state interfaces have to be created and transferred according
-   * to the hardware info passed in for the configuration.
+   * Old way of exporting the StateInterfaces. If a empty vector is returned then
+   * the on_export_state_interfaces() method is called. If a vector with StateInterfaces is returned
+   * then the exporting of the StateInterfaces is only done with this function and the ownership is
+   * transferred to the resource manager. The set_command(...), get_command(...), ..., can then not
+   * be used.
    *
    * Note the ownership over the state interfaces is transferred to the caller.
    *
    * \return vector of state interfaces
    */
-  virtual std::vector<StateInterface> export_state_interfaces() = 0;
+  [[deprecated(
+    "Replaced by vector<std::shared_ptr<StateInterface>> on_export_state_interfaces() method. "
+    "Exporting is "
+    "handled "
+    "by the Framework.")]] virtual std::vector<StateInterface>
+  export_state_interfaces()
+  {
+    // return empty vector by default. For backward compatibility we try calling
+    // export_state_interfaces() and only when empty vector is returned call
+    // on_export_state_interfaces()
+    return {};
+  }
+
+  /**
+   * Override this method to export custom StateInterfaces which are not defined in the URDF file.
+   * Those interfaces will be added to the unlisted_state_interfaces_ map.
+   *
+   *  Note method name is going to be changed to export_state_interfaces() as soon as the deprecated
+   * version is removed.
+   *
+   * \return vector of descriptions to the unlisted StateInterfaces
+   */
+  virtual std::vector<hardware_interface::InterfaceDescription>
+  export_state_interface_descriptions()
+  {
+    // return empty vector by default.
+    return {};
+  }
+
+  /**
+   * Default implementation for exporting the StateInterfaces. The StateInterfaces are created
+   * according to the InterfaceDescription. The memory accessed by the controllers and hardware is
+   * assigned here and resides in the actuator_interface.
+   *
+   * \return vector of shared pointers to the created and stored StateInterfaces
+   */
+  virtual std::vector<std::shared_ptr<StateInterface>> on_export_state_interfaces()
```

Comment:
```suggestion
  virtual std::vector<StateInterface::SharedPtr> on_export_state_interfaces()
```


Can we have the SharedPtr defined in the handle.hpp and reuse them. I think it is much cleaner and better this way.